spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/annotation/TaskListenerExecutor.java

```diff
@@ -0,0 +1,246 @@
+		this.failedTaskInstances = new HashMap<>();
+		initializeExecutor();
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;BeforeTask.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskStartup(TaskExecution taskExecution) {
+		executeTaskListener(taskExecution, beforeTaskMethods.keySet(), beforeTaskInstances);
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;AfterTask.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskEnd(TaskExecution taskExecution) {
+		executeTaskListener(taskExecution, afterTaskMethods.keySet(), afterTaskInstances);
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;FailedTask.
+	 * @param throwable that was not caught for the task execution.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskFailed(TaskExecution taskExecution, Throwable throwable) {
+		executeTaskListenerWithThrowable(taskExecution, throwable,
+				failedTaskMethods.keySet(),failedTaskInstances);
+	}
+
+	private void executeTaskListener(TaskExecution taskExecution, Set<Method> methods, Map<Method, Object> instances){
+		for (Method method : methods) {
+			try {
+				method.invoke(instances.get(method),taskExecution);
+			}
+			catch (IllegalAccessException e) {
+				throw new IllegalStateException("@BeforeTask and @AfterTask annotated methods must be public.");
+			}
+			catch (InvocationTargetException e) {
+				throw new IllegalStateException("Failed to process @BeforeTask or @AfterTask" +
+						"annotation because: ", e);
+			}
+			catch (IllegalArgumentException e){
+				throw new IllegalStateException("taskExecution parameter is required for @BeforeTask and @AfterTask annotated methods");
+			}
+		}
+	}
+
+	private void executeTaskListenerWithThrowable(TaskExecution taskExecution,
+			Throwable throwable, Set<Method> methods, Map<Method, Object> instances){
+		for (Method method : methods) {
+			try {
+				method.invoke(instances.get(method),taskExecution, throwable);
+			}
+			catch (IllegalAccessException e) {
+				throw new IllegalStateException("@FailedTask annotated methods must be public.");
+			}
+			catch (InvocationTargetException e) {
+				throw new IllegalStateException("Failed to process @FailedTask " +
+						"annotation because: ", e);
+			}
+			catch (IllegalArgumentException e){
+				throw new IllegalStateException("taskExecution and throwable parameters "
+						+ "are required for @FailedTask annotated methods");
+			}
+		}
+	}
+
+	private void initializeExecutor( ) {
+		ConfigurableListableBeanFactory factory = context.getBeanFactory();
+		for( String beanName : context.getBeanNamesForType(Object.class)) {
+
+			if (!ScopedProxyUtils.isScopedTarget(beanName)) {
+				Class<?> type = null;
+				try {
+					type = AutoProxyUtils.determineTargetClass(this.context.getBeanFactory(), beanName);
+				}
+				catch (Throwable ex) {
```

Comment:
In this case when determining the target class a lazy bean could thrown an exception.  The beans that utilize the annotations should not be lazy.  I can kick the message to info or warning.